### PR TITLE
Consider types assignable to open generics (#954)

### PR DESCRIPTION
**Pull request by mdonoughe**
I think I did this right, at least for BeAssignableTo. Base types and interfaces are searched for matching generic types.

Should BeDerivedFrom do the same thing but disallowing interfaces?

#954

## IMPORTANT 

* [x] The [Pull Request](https://help.github.com/articles/using-pull-requests) is targeted at the `master` branch.
* [ ] The code complies with the [Coding Guidelines for C# 3.0, 4.0 and 5.0](http://www.csharpcodingguidelines.com/).
* [x] The changes are covered by a new or existing set of unit tests which follow the Arrange-Act-Assert syntax such as is used [in this example](https://github.com/fluentassertions/fluentassertions/blob/daaf35b9b59b622c96d0c034e8972a020b2bee55/Tests/FluentAssertions.Shared.Specs/BasicEquivalencySpecs.cs#L33).
* [ ] If the contribution affects the documentation, please include your changes to [**documentation.md**](https://github.com/fluentassertions/fluentassertions/blob/master/docs/documentation.md) in this pull request so the documentation will appear on the [website](http://fluentassertions.com/documentation.html).

### CODE REVIEW

File: Src/FluentAssertions/Types/TypeAssertions.cs [hunk 1]
```diff
@@ -72,6 +72,45 @@ public new AndConstraint<TypeAssertions> BeAssignableTo<T>(string because = "",
             return BeAssignableTo(typeof(T), because, becauseArgs);
         }
 
+        bool CheckIfAssignableToOpenGeneric(Type definition)
```
Comment (dennisdoomen) [inline Src/FluentAssertions/Types/TypeAssertions.cs]: 🔧 Splitting the method in two, one for interfaces and one for classes might help read this code
🔧 Please move the method below the first use. We follow the order of execution ordering rule.

File: Src/FluentAssertions/Types/TypeAssertions.cs [hunk 2]
```diff
@@ -72,6 +72,45 @@ public new AndConstraint<TypeAssertions> BeAssignableTo<T>(string because = "",
...
+            // assignable to any generic type based on the given generic type definition.
+
+            if (definition.GetTypeInfo().IsInterface)
+            {
+                // check subject and its interfaces against definition
+                TypeInfo subjectInfo = Subject.GetTypeInfo();
+                if (subjectInfo.IsInterface && subjectInfo.IsGenericType &&
+                    subjectInfo.GetGenericTypeDefinition().IsSameOrEqualTo(definition))
+                {
+                    return true;
```
Comment (dennisdoomen) [inline Src/FluentAssertions/Types/TypeAssertions.cs]: 🙃 I personally find all these intermediate `return` statements a more difficult to read.

File: Src/FluentAssertions/Types/TypeAssertions.cs [hunk 3]
```diff
@@ -113,8 +162,18 @@ public new AndConstraint<TypeAssertions> NotBeAssignableTo<T>(string because = "
         /// <returns>An <see cref="AndConstraint{T}"/> which can be used to chain assertions.</returns>
         public new AndConstraint<TypeAssertions> NotBeAssignableTo(Type type, string because = "", params object[] becauseArgs)
         {
+            bool condition;
```
Comment (dennisdoomen) [inline Src/FluentAssertions/Types/TypeAssertions.cs]: 🔧 Maybe rename it to `isAssignable`?

File: Tests/Shared.Specs/TypeAssertionSpecs.cs [hunk 4]
```diff
@@ -484,6 +484,65 @@ public void When_an_unrelated_type_instance_it_fails_with_a_useful_message()
                 .WithMessage($"*{typeof(DummyImplementingClass)} to be assignable to {typeof(DateTime)}*failure message*");
         }
 
+        [Fact]
+        public void When_constructed_of_open_generic_it_succeeds()
```
Comment (dennisdoomen) [inline Tests/Shared.Specs/TypeAssertionSpecs.cs]: 🙃  We normally use the `When_...._it_should_....` format for all our tests. 
🤔 I don't get the _when constructed_. We're not constructing anything here, right?

File: Tests/Shared.Specs/TypeAssertionSpecs.cs [hunk 5]
```diff
@@ -484,6 +484,65 @@ public void When_an_unrelated_type_instance_it_fails_with_a_useful_message()
...
+        {
+            //-----------------------------------------------------------------------------------------------------------
+            // Arrange / Act / Assert
+            //-----------------------------------------------------------------------------------------------------------
+            typeof(IDummyInterface<IDummyInterface>).Should().BeAssignableTo(typeof(IDummyInterface<>));
+        }
+
+        [Fact]
+        public void When_implementation_of_open_generic_interface_it_succeeds()
```
Comment (dennisdoomen) [inline Tests/Shared.Specs/TypeAssertionSpecs.cs]: 🙃 It feels like the name is missing a verb or something.

Comment (mdonoughe) [reply Tests/Shared.Specs/TypeAssertionSpecs.cs]: I was trying to emulate the "When_its_own_type_instance_it_succeeds" pattern the other BeAssignableTo methods were following without making the names super long.

File: Src/FluentAssertions/Types/TypeAssertions.cs (hunk 2 above)
Comment (mdonoughe) [reply Src/FluentAssertions/Types/TypeAssertions.cs]: Is this better now?

**Comment by mdonoughe** [comment]
I've added in support for BeDerivedFrom too.

**Comment by mdonoughe** [comment]
BeOfType already had open generics.

Is there a convention for both `BeAssignableTo`s to use the same logic?

**Comment by jnyrup** [comment]
Yes, but `NotBeOfType` does not.

I'm not sure I understand your question.
If you're asking where to put shared logic, I would put it in `TypeExtensions.cs`.

**Comment by jnyrup** [comment]
Don't know if this is WIP, but `BeAssignableTo<T>` should probably call `BeAssignableTo(Type)` instead?

**Comment by mdonoughe** [comment]
It probably should, but it doesn't seem to make any difference since you can't pass open generics to methods like that. I assumed it did already.

**Comment by mdonoughe** [comment]
Actually should it? `BeAssignableTo<T>` casts the result to `T`, so for cases where it wouldn't match `BeAssignableTo` (shouldn't be any?) it would fail to cast even if it relied on `BeAssignableTo` to do the assertion.

**Comment by jnyrup** [comment]
It looks very good now and feature wise could be merge.
But I would like to have both a positive and a negative test for `BeAssignableTo`, `NotBeAssignableTo` and `NotBeOfType` from `ReferenceTypeAssertions` when the expectation is an open generic.

**Comment by dennisdoomen** [comment]
Maybe it's better to request changes then. It makes the state of the PR more visual.

**Pull request merged**